cdap-watchdog/src/main/java/io/cdap/cdap/logging/appender/LogAppender.java

```diff
@@ -123,38 +122,42 @@ private void addErrorCodeTags(ILoggingEvent event, Map<String, String> modifiabl
     }
   }
 
-  private void addErrorClassificationTags(ILoggingEvent event, Map<String, String> modifiableMDC) {
+  private void addErrorClassificationTags(ILoggingEvent event, Map<String, String> modifiableMdc) {
     if (event.getThrowableProxy() == null
         || !(event.getThrowableProxy() instanceof ThrowableProxy)) {
       return;
     }
     Throwable throwable = ((ThrowableProxy) (event.getThrowableProxy())).getThrowable();
-    while(throwable != null) {
-      if (throwable instanceof WrappedStageException) {
-        modifiableMDC.put(Logging.TAG_FAILED_STAGE,
-            ((WrappedStageException) throwable).getStageName());
+    while (throwable != null) {
+      if (throwable instanceof FailureDetailsProvider) {
+        FailureDetailsProvider provider = (FailureDetailsProvider) throwable;
+        modifiableMdc.put(Constants.Logging.TAG_FAILED_STAGE, provider.getFailureStage());
+        modifiableMdc.put(Constants.Logging.TAG_ERROR_CATEGORY,
+            provider.getErrorCategory().getErrorCategory());
+        modifiableMdc.put(Constants.Logging.TAG_ERROR_REASON, provider.getFailureReason());
+        modifiableMdc.put(Constants.Logging.TAG_ERROR_TYPE, provider.getErrorType().name());
       }
       if (throwable instanceof ProgramFailureException) {
-        modifiableMDC.put(Constants.Logging.TAG_ERROR_CATEGORY,
+        modifiableMdc.put(Constants.Logging.TAG_ERROR_CATEGORY,
             ((ProgramFailureException) throwable).getErrorCategory());
-        modifiableMDC.put(Constants.Logging.TAG_ERROR_REASON,
+        modifiableMdc.put(Constants.Logging.TAG_ERROR_REASON,
```

Comment:
I am wondering if we can make ProgramFailureException also implement FailureDetailsProvider. Most of the methods have same name, the one that is different on the first glance is getErrorReason vs getFailureReason, but we can rename it now while we introduce the interface

Reply:
ack, done.

Renamed `getFailureReason` to `getErrorReason` to make ProgramFailureException also implement FailureDetailsProvider.